File: skills/code-graph/tests/fixture/python/app/service.py
import os
import json as js

from .models import Repo, User


class Service:
    def __init__(self):
        self.repo = Repo()

    def run(self, u: User):
        self.repo.save(u)          # typed: inferred field type Repo
        os.path.join("a", "b")     # external namespace
        js.dumps(u)                # external alias -> must not hit models.dumps


def helper(x, u):
    x.save(u)                      # unknown receiver -> ambiguous lead at most


def make_repo() -> Repo:
    return Repo()


class Factory:
    def __init__(self):
        self.repo2 = make_repo()   # field typed from make_repo's return annotation

    def repo(self) -> Repo:
        return Repo()

    def run(self, u):
        r = make_repo()
        r.save(u)                  # typed via make_repo's return annotation
        self.repo().save(u)        # typed via the method's return annotation (call segment in the chain)
        self.repo2.save(u)         # typed via the field's call result

File: skills/code-graph/tests/fixture/js/src/service.js
import { Repo, load } from './repo.js';
import * as util from './util.js';
const fs = require('fs');
import './theme.css';                 // asset: neither an edge nor an unresolved import
import logo from './logo.svg?url';    // asset with a query
import { fmt2 } from '@fx/util';      // workspace package -> packages/util/src/index.ts (dist mapped to src)

export class Service {
  run() {
    const repo = new Repo();
    repo.save(1);        // typed: local declared via new Repo()
    load();              // free name from an import -> import
    util.fmt(2);         // namespace alias of a repo module -> import
    fs.readFile('x');    // external require -> unresolved, never util.readFile
    fmt2(3);             // workspace package function -> import
  }
}

export function loose(x) {
  x.save(2);             // unknown receiver -> ambiguous
}
